doc/developer/design/20231213_compute_operator_hydration_status_tracking.md

```diff
@@ -0,0 +1,192 @@
+In the future we should consider introducing additional regions to wrap cyclic dataflow parts, to improve the presentation of operator graphs we expose to users.
+
+### Hydration Status Logging
+
+Timely logging does not directly expose dataflow operator frontiers.
+While it does expose per-operator reachability information that can be used to derive frontiers, doing so requires accurate knowledge about the connectivity of the dataflow graph and complicated reasoning logic.
+To reduce complexity in the Compute code, we choose not to attempt this approach.
+
+Instead, we propose adding a hydration status logging operator at the end of each LIR region.
+These operators will observe frontier progress in their inputs and report hydration events via loggers connected to the `ComputeState`.
+There is already precedence for such operators, namely the ones that log import frontiers and arrangement sizes.
+
+### Reporting Introspection Data
+
+Existing replica introspection data is exposed through per-replica logging dataflows that export introspection arrangements.
+There are two issues with this approach:
+
+1. In multi-replica clusters, different replicas usually have different data in these arrangements.
+   As a result, introspection arrangements can only be queried by special replica-targeted queries.
+   To collect introspection data from all replicas, clients have to enumerate the existing replicas and issue a separate query for each, yielding bad UX.
+2. When a replica is slow to respond to queries, as if often the case during hydration, it will also be slow to respond to introspection queries.
+   This is particularly annoying when one wants to use introspection queries to monitor hydration progress.
+
+To avoid these issues, we propose reporting operator hydration information to the compute controller through `ComputeResponse`s instead.
+The compute controller can then write this information into storage-managed collections and expose it to SQL clients that way.
+This solves issue (1) as introspection information from all replicas ends up in the same collection and can be retrieved with a single query.
+Issue (2) is mostly solved because the time logging events spend on the replica is reduced.
+It is still the case that when a replica is completely unresponsive, it likely won't schedule the hydration logging operators, so no new introspection data will make it to the compute controller either.
+However, even in these (hopefully rare) cases introspection data previously written can still be queried.
+
+### SQL-level Changes
+
+This design proposes adding two new relations in which operator hydration status information is exposed:
+
+```
+mz_internal.mz_compute_operator_hydration_statuses_per_worker {
+    object_id: text,
+    physical_plan_id: uint8,
```

Comment:
Is this a uint8 because they are usizes in the implementation? Fine if so, otherwise I generally have preference to avoid the uint types in favor of signed in sql interfaces because it's a non-standard type.

Reply:
Mostly because IDs can't be negative and my programmer brain insists on using signed integers only for things that can be negative. But your argument checks out, I'll change this.

Reply:
Actually, I checked our catalog docs and it looks like `uint8` is used way more often than `int8` throughout both `mz_catalog` and `mz_internal`. Shouldn't we stick with `uint8` then, for consistency?